examples/rp/src/bin/pio_uart.rs

```diff
@@ -0,0 +1,431 @@
+    static USB_CHANNEL_TX: Channel<ThreadModeRawMutex, u8, 20> = Channel::<ThreadModeRawMutex, u8, 20>::new();
+    let mut usb_channel_tx_send = USB_CHANNEL_TX.sender();
+    let mut usb_channel_tx_recv = USB_CHANNEL_TX.receiver();
+
+    static UART_CHANNEL_TX: Channel<ThreadModeRawMutex, u8, 20> = Channel::<ThreadModeRawMutex, u8, 20>::new();
+    let mut uart_channel_tx_send = UART_CHANNEL_TX.sender();
+    let mut uart_channel_tx_recv = UART_CHANNEL_TX.receiver();
+
+    let (mut usb_tx, mut usb_rx) = class.split();
+
+    // Read + write from USB
+    let usb_future = async {
+        loop {
+            info!("Wait for USB connection");
+            usb_rx.wait_connection().await;
+            info!("Connected");
+            let _ = join(
+                usb_read(&mut usb_rx, &mut uart_channel_tx_send),
+                usb_write(&mut usb_tx, &mut usb_channel_tx_recv),
+            )
+            .await;
+            info!("Disconnected");
+        }
+    };
+
+    // Read + write from UART
+    let uart_future = async {
+        loop {
+            let _ = join(
+                uart_read(&mut uart_rx, &mut usb_channel_tx_send),
+                uart_write(&mut uart_tx, &mut uart_channel_tx_recv),
+            )
+            .await;
+        }
+    };
+
+    // Run everything concurrently.
+    // If we had made everything `'static` above instead, we could do this using separate tasks instead.
+    join3(usb_fut, usb_future, uart_future).await;
+}
+
+struct Disconnected {}
+
+impl From<EndpointError> for Disconnected {
+    fn from(val: EndpointError) -> Self {
+        match val {
+            EndpointError::BufferOverflow => panic!("Buffer overflow"),
+            EndpointError::Disabled => Disconnected {},
+        }
+    }
+}
+
+/// Read from the USB and write it to the UART TX send channel
+async fn usb_read<'d, T: Instance + 'd>(
+    usb_rx: &mut Receiver<'d, Driver<'d, T>>,
+    uart_tx_send: &mut embassy_sync::channel::Sender<'d, ThreadModeRawMutex, u8, 20>,
+) -> Result<(), Disconnected> {
+    let mut buf = [0; 64];
+    loop {
+        let n = usb_rx.read_packet(&mut buf).await?;
+        let data = &buf[..n];
+        trace!("USB IN: {:x}", data);
+        for byte in data {
+            uart_tx_send.send(*byte).await;
+        }
+    }
+}
+
+/// Read from the USB TX receive channel and write it to the USB
+async fn usb_write<'d, T: Instance + 'd>(
+    usb_tx: &mut Sender<'d, Driver<'d, T>>,
+    usb_tx_recv: &mut embassy_sync::channel::Receiver<'d, ThreadModeRawMutex, u8, 20>,
+) -> Result<(), Disconnected> {
+    loop {
+        let n = usb_tx_recv.recv().await;
+        let data = [n];
+        trace!("USB OUT: {:x}", data);
+        usb_tx.write_packet(&data).await?;
+    }
+}
+
+/// Read from the UART and write it to the USB TX send channel
+async fn uart_read<'a>(
+    uart_rx: &mut PioUartRx<'a>,
+    usb_tx_send: &mut embassy_sync::channel::Sender<'a, ThreadModeRawMutex, u8, 20>,
+) -> Result<(), Disconnected> {
```

Comment:
if `uart_read`/`uart_write` never return, it'd be more idiomatic to make them `-> !`

Reply:
Fixed